[SPARK-17062][MESOS] add conf option to mesos dispatcher

skonto opened the pull request:
## What changes were proposed in this pull request?

Adds --conf option to set spark configuration properties in mesos dispacther.
Properties provided with --conf take precedence over properties within the properties file.
The reason for this PR is that for simple configuration or testing purposes we need to provide a property file (ideally a shared one for a cluster) even if we just provide a single property.
## How was this patch tested?

Manually tested.

Commit:
add conf option to mesos dispatcher

skonto commented:
@mgummelt pls review.

SparkQA commented:
**[Test build #63790 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63790/consoleFull)** for PR 14650 at commit [`db42e31`](https://github.com/apache/spark/commit/db42e31e2390bfa5dd59e0e4eed6352d171e29de).

SparkQA commented:
**[Test build #63790 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63790/consoleFull)** for PR 14650 at commit [`db42e31`](https://github.com/apache/spark/commit/db42e31e2390bfa5dd59e0e4eed6352d171e29de).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63790/
Test PASSed.

rxin commented:
Does the existing spark-submit not work on Mesos?

skonto commented:
I think not.
MesosDispacther uses [spark-daemon.sh](https://github.com/apache/spark/blob/master/sbin/start-mesos-dispatcher.sh#L44) which uses [class](https://github.com/apache/spark/blob/master/sbin/spark-daemon.sh#L191) not submit:
It also creates the sparkConfig from [scratch](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcher.scala#L97)/

skonto commented:
@srowen @mgummelt  what do you think?

srowen commented:
Seems OK to me, to the limits of my understanding, and given the logic of https://github.com/apache/spark/pull/14650#issuecomment-240057336

skonto commented:
@mgummelt?

skonto commented:
@srowen could you merge it pls?

mgummelt commented:
I'm generally fine with this, though one downside is that it introduces a consistency with other daemon classes such as Master.scala, which only accepts a properties file.  Maybe we should make a JIRA to add this to the other classes.

Not that Spark configuration has any sort of sane consistency to be upheld.  There are ~10 ways of setting config properties.  It's one of the most confusing things about operations.

@srowen Do you know why even have a separate set of Spark config properties, rather than just using Java System properties?  I know you can load Spark conf from Java properties, but you can also load them via `--conf` (and only in spark-submit), which seems like an unnecessary nonstandard interface.

mgummelt commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -73,6 +82,10 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
       parse(tail)
 
+    case ("--conf") :: value :: tail =>
+      cliSparkConfig += value
```
Let's make this consistent with `SparkSubmitArguments`, and parse these into a HashMap here.

skonto replied:
I prefer it this way seems natural to me if you follow the logic

mgummelt commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,26 @@ private[spark] object Utils extends Logging {
+      k.startsWith("spark.")
+    }.foreach { case (k, v) =>
+      conf.set(k, v)
+      sys.props.getOrElseUpdate(k, v)
```
Why set them as System properties as well?

skonto replied:
For consistency if you check a few lines above with other methods which do the same.

mgummelt commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -102,7 +115,9 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
         "  --properties-file FILE  Path to a custom Spark properties file.\n" +
-        "                          Default is conf/spark-defaults.conf.")
+        "                          Default is conf/spark-defaults.conf \n" +
+        "  --conf PROP=VALUE       Arbitrary Spark configuration property.\n" +
```
The work "arbitrary" is unnecessary.  Not a blocker.

skonto replied:
I just copied the help output of spark-submit. For consistency reasons lets keep it that way.

mgummelt commented on core/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -102,7 +115,9 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
-        "                          Default is conf/spark-defaults.conf.")
+        "                          Default is conf/spark-defaults.conf \n" +
+        "  --conf PROP=VALUE       Arbitrary Spark configuration property.\n" +
+        "                          Takes precedence over defined properties in properties-file.")
```
The word "defined" is unnecessary

mgummelt commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,26 @@ private[spark] object Utils extends Logging {
+  }
+
+  /** Loads properties from a String */
+  def loadPropertiesFromString(input: String) : Properties = {
```
I don't think we need this.  We can parse the input into either a Properties object or a HashMap in the above file.

skonto replied:
Yes there are different ways you can do the same thing.

skonto replied:
The reason I did it that way is to avoid splitting the "property=value"  thing mysellf in the first place... I let the load method to do that.

skonto replied:
If this is a blocker let me know.

mgummelt replied:
My only issue is that this is inconsistent with how SparkSubmitArguments parses {{--conf}}.  So if you prefer this approach, let's refactor SparkSubmitArguments to use this function as well.

skonto replied:
Ok I will try to use the same function either the one in SparkSubmitArguments or load. The error message must remain the same one way or another.

mgummelt commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,26 @@ private[spark] object Utils extends Logging {
+   * Updates Spark config with properties from a set of Properties.
+   * Provided properties have the highest priority.
+   */
+  def updateSparkConfigFromProperties(conf: SparkConf, properties: Properties): Unit = {
```
I think you can fold this into `loadFromSystemProperties` in `SparkConf`

skonto replied:
Yes with the exception of this line: sys.props.getOrElseUpdate(k, v) for which again as I said I am not sure.

mgummelt replied:
I'm pretty sure we don't need to set system properties.  They're only used to set SparkConf anyway, I think.

skonto replied:
Ok I will remove it and use that function in SparkConf.

srowen commented:
Responding to https://github.com/apache/spark/pull/14650#issuecomment-241095780 , IMHO it's best if configuration never uses environment variables or system properties. They're global to a process and that eventually ends in some tears, despite its convenience. Hence I personally support standardizing towards explicit configuration. 

In practice some env variables are used here and there for legacy reasons. And yes you can set props via system properties. I think being consistent is OK, right or wrong, but would generally move away from env / sys properties if given the choice.

skonto commented:
WIP

skonto committed:
Merge branch 'master' into dipatcher_conf

SparkQA commented:
**[Test build #65311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65311/consoleFull)** for PR 14650 at commit [`3532f5d`](https://github.com/apache/spark/commit/3532f5db94e8bcd85e87c9a87fc4ce4d42a50655).

SparkQA commented:
**[Test build #65311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65311/consoleFull)** for PR 14650 at commit [`3532f5d`](https://github.com/apache/spark/commit/3532f5db94e8bcd85e87c9a87fc4ce4d42a50655).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65311/
Test FAILed.

skonto committed:
refactor

SparkQA commented:
**[Test build #65333 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65333/consoleFull)** for PR 14650 at commit [`497e517`](https://github.com/apache/spark/commit/497e517d9e26970ab0990490186fc12abdbbef46).

skonto commented:
@mgummelt @srowen I refactored a bit MesosSchedulerDispatcherArguments to follow the SparkSubmit pattern. Re-used some code there. Added several tests.

SparkQA commented:
**[Test build #65333 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65333/consoleFull)** for PR 14650 at commit [`497e517`](https://github.com/apache/spark/commit/497e517d9e26970ab0990490186fc12abdbbef46).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait SparkConfigParseUtils `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65333/
Test FAILed.

skonto commented:
WIP fixing test...

skonto committed:
fix test

SparkQA commented:
**[Test build #65358 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65358/consoleFull)** for PR 14650 at commit [`342d8d3`](https://github.com/apache/spark/commit/342d8d32de5290600213758ce2c21dcf1f0c5b69).

SparkQA commented:
**[Test build #65358 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65358/consoleFull)** for PR 14650 at commit [`342d8d3`](https://github.com/apache/spark/commit/342d8d32de5290600213758ce2c21dcf1f0c5b69).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65358/
Test PASSed.

skonto commented:
@srowen @mgummelt pls review...

skonto commented:
@srowen what do you think?

srowen commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -40,12 +40,12 @@ import org.apache.ivy.plugins.matcher.GlobPatternMatcher
 
-import org.apache.spark.{SPARK_REVISION, SPARK_VERSION, SparkException, SparkUserAppException}
 import org.apache.spark.{SPARK_BRANCH, SPARK_BUILD_DATE, SPARK_BUILD_USER, SPARK_REPO_URL}
+import org.apache.spark.{SPARK_REVISION, SPARK_VERSION, SparkException, SparkUserAppException}
```
This is trivial, but I think at this point a "_" import is fine and called for. It's split over two imports now.

skonto replied:
Ok cool It was there before my PR but I will fix it no problem.

skonto replied:
Fixed

srowen commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -27,28 +27,18 @@ import org.scalatest.{BeforeAndAfterEach, Matchers}
 import org.scalatest.time.SpanSugar._
 
-import org.apache.spark._
+import org.apache.spark.{SparkFunSuite, _}
```
Not needed? it already imports everything

skonto replied:
Correct will fix.

srowen commented on mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -73,37 +94,55 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
       }
 
-    case _ =>
+    case value@_ =>
```
`case value =>` or maybe I misunderstand why it's written this way

skonto replied:
Its the same thing case value => less verbose. Will change it.

skonto replied:
fixed.

srowen commented on core/src/main/scala/org/apache/spark/util/Executable.scala:
```diff
@@ -0,0 +1,25 @@
+/*
+ * Adds main method to a class or object
+ */
+private[spark] trait Executable {
```
This trait feels funny. I didn't see what it is needed for though. It's used in tests only, but isn't used in a place where you need to abstract over both implementations?

skonto replied:
I needed to mark specific classes with main support, I could have used a different implementation to find out about it like duck typing, or reflection etc but these are either not permitted (eg. scala style) or are too complex. It is not bad to tag a class that way. Maybe sound redundant even funny (I would not call that way) but it serves a purpose to make things easily testable.

skonto replied:
Ok so in the code there is no reason as you said. In the test code I tried to re-use this function [here]:
https://github.com/skonto/spark/blob/d40b928aba0f44c5b3397d61d4450489e5d387c6/core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala#L61
and I need to pass an object for testing and that object needs to implement some specific functions.
Spark project does not allow duck typing and I thought reflection is too much. Anyway I can eliminate it with reflection. I could have avoided in the first place but not reusing the function but I thought I could improve test code too...

skonto replied:
Anyway I can eliminate it with reflection, it is test code anyway.

vanzin replied:
Agree that this trait seems unnecessary. For the test, you can use two arguments: a "SparkConfigParsingUtils" and a second one that's a main function (`main: Array[String] => Unit`).

vanzin replied:
This trait still exists and still looks out of place. It could be merged with `CommandLineUtils`, which then could be called something like `SparkEntryPoint` (although `CommandLineUtils` is also probably fine).

srowen commented on mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -18,23 +18,43 @@
 
+  // scalastyle:on println
   propertiesFile = Utils.loadDefaultSparkProperties(conf, propertiesFile)
+  Utils.updateSparkConfigFromProperties(conf, confProperties)
```
This seems like it differs from Mesos to non-Mesos implementations. Here you can update props with --conf but wouldn't this be identical elsewhere? I may have missed the purpose.

skonto replied:
Logic differs wrt to others (I would have to refactor everything otherwise which is not wise given code size). I just added support for the --conf option, that is the same across implementations, and re-used common code for parsing stuff.

skonto replied:
fixed.

skonto replied:
Non mesos code does not support --conf or they dont care in all cases about spark. confs.
Check here:
https://github.com/apache/spark/blob/master/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMasterArguments.scala
and here:
https://github.com/apache/spark/blob/master/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala#L753

If you compare the same functionality (the one about loading the properties file) with the MesosClusterDispatcherArguments file, things are completely different, before my PR.

https://github.com/apache/spark/blob/master/mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala#L37

So I dont expect parallel similarity since for example in this case ApplicationMaster does not need to get the spark configuration in a Spark Config.

srowen reviewed (changes requested):
CC @vanzin because this is touching spark submit.

SparkQA commented:
**[Test build #65594 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65594/consoleFull)** for PR 14650 at commit [`d40b928`](https://github.com/apache/spark/commit/d40b928aba0f44c5b3397d61d4450489e5d387c6).

SparkQA commented:
**[Test build #65594 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/65594/consoleFull)** for PR 14650 at commit [`d40b928`](https://github.com/apache/spark/commit/d40b928aba0f44c5b3397d61d4450489e5d387c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/65594/
Test PASSed.

srowen commented:
OK I understand more about why the code is structured this way. Let me back up another step before coming back to the details.

It feels like this change is trying to unify `MesosClusterDispatcher` and `SparkSubmit`. That doesn't feel like the role of `MesosClusterDispatcher`. It seems like a peer of things like the `Master` class for standalone. 

I wonder why the `YARNAllocator` or `Master` doesn't have a `--conf` argument ... they seem to receive config via `SparkConf` which in turn would have been set up by things like `spark-submit`.

That may be why refactoring across these two is hard for me to process here.

Calling in @vanzin if possible for a comment.

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,20 @@ private[spark] object Utils extends Logging {
+   * Updates Spark config with properties from a set of Properties.
+   * Provided properties have the highest priority.
+   */
+  def updateSparkConfigFromProperties(conf: SparkConf, properties: Map[String, String])
```
If you're gonna break method declarations into multiple lines, follow the appropriate style:
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Indentation

skonto replied:
Missed that.

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,20 @@ private[spark] object Utils extends Logging {
+  def updateSparkConfigFromProperties(conf: SparkConf, properties: Map[String, String])
+    : Unit = {
+    properties.filter { case (k, v) =>
+      k.startsWith("spark.")
```
You can move this to the `foreach` body.

On a side note, this whole method feels a little funny, but I have to see how it's used.

skonto replied:
I just re-used the code found [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/Utils.scala#L1952).

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -88,6 +82,22 @@ class SparkSubmitSuite
+
+  override def beforeEach() {
+    super.beforeEach()
+    System.setProperty("spark.testing", "true")
```
You shouldn't need to do this. Otherwise it would mean a bug in `ResetSystemProperties`.

skonto replied:
It is not my code...

vanzin commented on mesos/src/main/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArguments.scala:
```diff
@@ -73,37 +94,55 @@ private[mesos] class MesosClusterDispatcherArguments(args: Array[String], conf:
-    case _ =>
+    case value =>
+      // scalastyle:off println
+      MesosClusterDispatcher.printStream.println(s"Unrecognized option: '${value.head}'")
```
Does `import MesosClusterDispatcher._` work to get these methods? Kinda noisy to see the class name repeated everywhere.

skonto replied:
I guess it is noisy too here: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala

Counted 34 appearances of SparkSubmit. I just folloed the same idea I thought since it was previously reviewed I should do the same, guessing it was for clarity reasons.

vanzin commented on core/src/main/scala/org/apache/spark/util/SparkConfigParseUtils.scala:
```diff
@@ -0,0 +1,48 @@
+ * Facilitates the use of common error messages when spark conf parsing fails.
+ *
+ */
+private[spark] trait SparkConfigParseUtils {
```
I'm not really in love with the name of this class. Not only it's awkward, but it doesn't reflect the actual functionality, since it has more than just utility methods for parsing configs from the command line.

I'd call it something more generic like `CommandLineUtils`, saying that it contains basic command line parsing functionality and methods to parse some common Spark CLI options.

skonto replied:
It is always a matter of taste... but anyway I can make it more generic np.

vanzin commented on core/src/main/scala/org/apache/spark/util/SparkConfigParseUtils.scala:
```diff
@@ -0,0 +1,48 @@
+  // Exposed for testing
+  private[spark] var exitFn: Int => Unit = (exitCode: Int) => System.exit(exitCode)
+  private[spark] var printStream: PrintStream = System.err
+  private[spark] def printWarning(str: String): Unit = printStream.println("Warning: " + str)
```
Could you add empty lines between `def`s?

skonto replied:
Sure.

vanzin commented on core/src/main/scala/org/apache/spark/util/SparkConfigParseUtils.scala:
```diff
@@ -0,0 +1,48 @@
+    exitFn(1)
+  }
+
+  private[spark] def parseSparkConfProperty(pair: String, update: (String, String) => Unit)
```
This signature is super awkward. I'd suggest two alternatives:

- return the parsed (key, value) pair and let the caller decide what to do
- move all config parsing into this trait and let it handle collecting the config values.

The latter could also include a method for handling `--properties-file`, and a way for getting the "final" Spark configuration. This functionality is kind of a pre-requisite for SPARK-12344, btw, which basically requires doing this for all processes that use "spark-class".

skonto replied:
The name just states that it parses a spark configuration property. I dont see it awkward. 
Anyway I will go with the first option.

skonto commented:
WIP

skonto committed:
fixes

skonto commented:
@vanzin & @srowen pls review.

skonto commented:
jenkins please test

SparkQA commented:
**[Test build #67066 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67066/consoleFull)** for PR 14650 at commit [`c322c27`](https://github.com/apache/spark/commit/c322c276ecbed352ea4dbf32df5b6f0d1f2c4347).

SparkQA commented:
**[Test build #67066 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67066/consoleFull)** for PR 14650 at commit [`c322c27`](https://github.com/apache/spark/commit/c322c276ecbed352ea4dbf32df5b6f0d1f2c4347).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67066/
Test PASSed.

skonto commented:
@mgummelt ?

vanzin commented on core/src/main/scala/org/apache/spark/util/CommandLineUtils.scala:
```diff
@@ -0,0 +1,49 @@
+    exitFn(1)
+  }
+
+  private[spark] def parseSparkConfProperty(pair: String): Either[Unit, (String, String)] = {
```
Returning `Either` here is weird. It should return `(String, String)`. I know you did this because of the error handling case, but that's better handled by:

```
printErrorAndExit(s"Spark config without '=': $pair")
throw new SparkException(s"Spark config without '=': $pair")
```

That way if `exitFn` doesn't really exit, you get the exception. Hopefully only tests override `exitFn` and those shouldn't really trigger the exception, so it's there just to make the compiler happy.

skonto replied:
ok...

skonto replied:
Ok cool I was not aware if I can thrown a SparkException at that point... specifically if it is alligned with the rest of the code base thats why I followed that weird path to make also the compiler happy with the return type...

skonto replied:
Ok I fixed that although throw new SparkException(s"Spark config without '=': $pair") is a bit redundant, it is there only to make compiler happy which is again a bit strange to me, anyway.

vanzin commented on core/src/main/scala/org/apache/spark/util/CommandLineUtils.scala:
```diff
@@ -0,0 +1,49 @@
+ */
+private[spark] trait CommandLineUtils {
+
+  // scalastyle:off println
```
nit: you're missing `// scalastyle:on println`

skonto replied:
yeah..

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -412,10 +412,9 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
-        }
+        SparkSubmit.parseSparkConfProperty(value)
+          .right
+          .foreach{case (k, v) => sparkProperties(k) = v}
```
nit: space before and after `{`

skonto replied:
ok

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,22 @@ private[spark] object Utils extends Logging {
+  def updateSparkConfigFromProperties(
+      conf: SparkConf,
+      properties: Map[String, String])
+    : Unit = {
```
nit: move to previous line

vanzin commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -65,16 +55,20 @@ class SparkSubmitSuite
+      input: Array[String],
+      searchString: String,
+      mainObject: Executable with CommandLineUtils = SparkSubmit)
+    : Unit = {
```
nit: move to previous line

vanzin commented on mesos/src/test/scala/org/apache/spark/deploy/mesos/MesosClusterDispatcherArgumentsSuite.scala:
```diff
@@ -0,0 +1,63 @@
+import org.apache.spark.deploy.TestPrematureExit
+
+class MesosClusterDispatcherArgumentsSuite extends SparkFunSuite
+  with TestPrematureExit{
```
nit: space before '{'

skonto replied:
ok

vanzin reviewed (commented):
I need to do a second pass (have to run now), but I really would like that `Executable` trait to go away.

skonto commented:
@vanzin ok I will update the PR asap.

skonto commented:
WIP

skonto commented:
@vanzin pls review.

SparkQA commented:
**[Test build #68714 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68714/consoleFull)** for PR 14650 at commit [`19dbd73`](https://github.com/apache/spark/commit/19dbd7330d82be1b465a5a905df5db6f5e6b326d).

SparkQA commented:
**[Test build #68714 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68714/consoleFull)** for PR 14650 at commit [`19dbd73`](https://github.com/apache/spark/commit/19dbd7330d82be1b465a5a905df5db6f5e6b326d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68714/
Test FAILed.

skonto commented:
jenkins please test

SparkQA commented:
**[Test build #68715 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68715/consoleFull)** for PR 14650 at commit [`19dbd73`](https://github.com/apache/spark/commit/19dbd7330d82be1b465a5a905df5db6f5e6b326d).

SparkQA commented:
**[Test build #68715 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68715/consoleFull)** for PR 14650 at commit [`19dbd73`](https://github.com/apache/spark/commit/19dbd7330d82be1b465a5a905df5db6f5e6b326d).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68715/
Test FAILed.

skonto commented:
I get some strange error with the jenkins build any idea?

```[error] (mesos/compile:compileIncremental) Compilation failed
[error] (core/test:compileIncremental) Compilation failed
[error] Total time: 342 s, completed Nov 16, 2016 4:37:45 AM
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/build/sbt -Phadoop-2.3 -Pmesos -Pkinesis-asl -Pyarn -Phive-thriftserver -Phive test:package streaming-kafka-0-8-assembly/assembly streaming-flume-assembly/assembly streaming-kinesis-asl-assembly/assembly ; received return code 1
Attempting to post to Github...
 > Post successful.
Build step 'Execute shell' marked build as failure
Archiving artifacts
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68715/
Test FAILed.
Finished: FAILURE

SparkQA commented:
**[Test build #68718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68718/consoleFull)** for PR 14650 at commit [`cfadc06`](https://github.com/apache/spark/commit/cfadc06bb144d0fbf98c3667ae0f85e4502d771b).

SparkQA commented:
**[Test build #68718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68718/consoleFull)** for PR 14650 at commit [`cfadc06`](https://github.com/apache/spark/commit/cfadc06bb144d0fbf98c3667ae0f85e4502d771b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68718/
Test PASSed.

skonto commented:
@srowen could I get a merge please?

rxin commented:
@vanzin did you do your 2nd pass?

vanzin commented:
Not yet, I'll try to find some time today.

vanzin commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -412,10 +412,8 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
-          case Seq(k, v) => sparkProperties(k) = v
-          case _ => SparkSubmit.printErrorAndExit(s"Spark config without '=': $value")
-        }
+        val pair = SparkSubmit.parseSparkConfProperty(value)
```
nit: `val (confName, confValue) = SparkSubmit.parseSparkConfProperty(value)`

skonto replied:
ok.

vanzin commented on core/src/main/scala/org/apache/spark/util/CommandLineUtils.scala:
```diff
@@ -0,0 +1,57 @@
+ */
+private[spark] trait CommandLineUtils {
+
+  // scalastyle:on println
```
nit: you don't need this

skonto replied:
ok

vanzin commented on core/src/main/scala/org/apache/spark/util/CommandLineUtils.scala:
```diff
@@ -0,0 +1,57 @@
+        throw new SparkException(s"Spark config without '=': $pair")
+    }
+  }
+  // scalastyle:on println
```
nit: you can move this after `printErrorAndExit`.

skonto replied:
ok

vanzin commented on core/src/main/scala/org/apache/spark/util/CommandLineUtils.scala:
```diff
@@ -0,0 +1,57 @@
+
+  // scalastyle:off println
+  private[spark] def printWarning(str: String): Unit = printStream.println("Warning: " + str)
+
```
nit: remove extra empty line

skonto replied:
ok.

skonto replied:
removed the empty line it still shows one.

vanzin commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -1962,6 +1962,21 @@ private[spark] object Utils extends Logging {
+      k.startsWith("spark.")
+    }.foreach { case (k, v) =>
+      conf.set(k, v)
+      sys.props.getOrElseUpdate(k, v)
```
Why is this method updating system properties? The name nor the scaladoc mention that fact, and this is a pretty important thing the method does. Not only that but the behavior is different from what it does to the SparkConf (set if not yet set, instead of always set).

skonto replied:
loadDefaultSparkProperties method a few lines above does that so I thought this is a common practice to load stuff there because they might be reused elsewhere. I can remove it for sure.

vanzin reviewed (commented):
I'm not sure `updateSparkConfigFromProperties` should be touching system properties. We really should avoid adding more code that messes with those.

Other than that, looks ok.

SparkQA commented:
**[Test build #68805 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68805/consoleFull)** for PR 14650 at commit [`946202d`](https://github.com/apache/spark/commit/946202de791f39529ef0aa05cadc0c65c1cd3f70).

skonto commented:
@vanzin I updated the PR. Thank you for the review.

SparkQA commented:
**[Test build #68805 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/68805/consoleFull)** for PR 14650 at commit [`946202d`](https://github.com/apache/spark/commit/946202de791f39529ef0aa05cadc0c65c1cd3f70).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/68805/
Test PASSed.

skonto commented:
@rxin @srowen could I get merge pls if there are no other issues?

srowen commented:
Since @vanzin has looked at this most closely I'd defer to him for a merge when you both think it's ready.

vanzin commented:
LGTM. There are a few style nits that I'll fix during the merge to avoid another round.

Merging to master.

skonto commented:
@vanzin if there is anything to fix I can do it if you want I can do another parse to check it against the style guide. For all the style nit stuff I guess sometimes is a matter of taste...

vanzin commented:
This is already merged. Maybe there's something wrong with the apache / github sync.

https://git-wip-us.apache.org/repos/asf?p=spark.git;a=commit;h=ea77c81ec0db27ea4709f71dc080d00167505a7d

rxin commented:
Guys this broke the build. Please don't use Option.contains in the future. It is both confusing and not working for Scala 2.10.

vanzin commented:
I see you already pushed a fix, thanks.
